src/libPMacc/include/particles/ParticlesBase.kernel

```diff
@@ -689,97 +689,200 @@ struct KernelDeleteParticles
-                    auto parSrc = (frame[threadIdx.x]);
-                    assign( parDest, parSrc );
-                    parSrc[multiMask_] = 0;
+                    if ( frame[ linearIdx ][ multiMask_ ] == 1 )
+                    {
+                        bashIdxCtx[ idx ] = nvidia::atomicAllInc( &numBashedParticles );
+                    }
                 }
+            );
+            __syncthreads( );
+
+            if( numBashedParticles > 0 )
+            {
+
+                onlyMaster(
+                    [&](
+                        uint32_t const,
+                        uint32_t const
+                    )
+                    {
+                        tmpBorder = border.pushN(
+                            numBashedParticles,
+                            // Compute the target supercell depending on the exchangeType
+                            DataSpaceOperations< dim >::reduce(
+                                superCellIdx,
+                                mapper.getExchangeType( )
+                            )
+                        );
+                        if( tmpBorder.getSize( ) < numBashedParticles )
+                            hasMemory = false;
+                    }
+                );
+
                 __syncthreads( );
 
-                if ( threadIdx.x == 0 && hasMemory )
-                {
-                    //always remove the last frame
-                    frame = getPreviousFrameAndRemoveLastFrame( frame, pb, superCellIdx );
-                }
+                forEachParticle(
+                    [&](
+                        uint32_t const linearIdx,
+                        uint32_t const idx
+                    )
+                    {
+                        if( bashIdxCtx[ idx ] != INV_LOC_IDX && bashIdxCtx[ idx ] < tmpBorder.getSize( ) )
+                        {
+                            auto parDest = tmpBorder[ bashIdxCtx[ idx ] ][ 0 ];
+                            auto parSrc = frame[ linearIdx ];
+                            assign( parDest, parSrc );
+                            parSrc[ multiMask_ ] = 0;
+                        }
+                    }
+                );
+                __syncthreads( );
             }
-            else
-            {
-                //if we had no particles to copy than we are the last and only frame
-                if ( threadIdx.x == 0 )
+
+            onlyMaster(
+                [&](
+                    uint32_t const,
+                    uint32_t const
+                )
                 {
-                    frame = getPreviousFrameAndRemoveLastFrame( frame, pb, superCellIdx );
+                    /* do not remove the frame if we had not enough memory for
+                     * all particles in the tmpBorder buffer
+                     */
+                    if ( hasMemory )
+                        frame = getPreviousFrameAndRemoveLastFrame( frame, pb, superCellIdx );
```

Comment:
`tmpBorder` can only have <= `numBashedParticles` elements and is only valid for one round (while loop).
I realized that `hasMemory` is bad named because the flag describe if it was possible to copy all particles to the border buffer.

I will rename `hasMemory` and will add better documentation to tmpBorder.